lib/compiler-llvm/src/config.rs

```diff
@@ -179,16 +194,47 @@ impl LLVM {
 
         let target_triple = self.target_triple(&target);
         let llvm_target = InkwellTarget::from_triple(&target_triple).unwrap();
-        llvm_target
+        let llvm_target_machine = llvm_target
             .create_target_machine(
                 &target_triple,
-                "generic",
-                &llvm_cpu_features,
+                match triple.architecture {
+                    Architecture::Riscv64(_) => "generic-rv64",
+                    _ => "generic",
+                },
+                match triple.architecture {
+                    Architecture::Riscv64(_) => "+m,+a,+c,+d,+f",
+                    _ => &llvm_cpu_features,
+                },
                 self.opt_level,
                 self.reloc_mode(),
-                self.code_model(),
+                match triple.architecture {
+                    Architecture::Riscv64(_) => CodeModel::Medium,
+                    _ => self.code_model(),
+                },
             )
-            .unwrap()
+            .unwrap();
+
+        if let Architecture::Riscv64(_) = triple.architecture {
+            // TODO: totally non-portable way to change ABI
+            unsafe {
+                pub struct MyTargetMachine {
+                    pub target_machine: *const u8,
+                }
+
+                let my_target_machine: MyTargetMachine = std::mem::transmute(llvm_target_machine);
+
+                *((my_target_machine.target_machine as *mut u8).offset(0x410) as *mut u64) = 5;
+                std::ptr::copy_nonoverlapping(
+                    "lp64d\0".as_ptr(),
+                    (my_target_machine.target_machine as *mut u8).offset(0x418),
+                    6,
+                );
+
+                std::mem::transmute(my_target_machine)
+            }
```

Comment:
This will need to be in llvm-sys or inkwell, right?

Reply:
Yes, but LLVM C API to change ABI doesn't seem to exist, so it should also be added.

Reply:
Thanks for the context.
Do you think we can create an issue on the LLVM project and add an extra comment in this source with a reference to the URL that new LLVM issue?

Reply:
Sorry for the late reply as I have been busy with my day job😢 
I checked LLVM again, and I found that it seems by this change the correct ABI will be specified in LLVM 15, so I will try it.
https://github.com/llvm/llvm-project/commit/4e115b7d881136947c083e12f62010bc6b1d3f00